### PR TITLE
Add embedding_dense_backward op

### PR DESCRIPTION
**Pull request by xytintel**
Add embedding_dense_backward op.

### CODE REVIEW

File: src/aten/Embedding.cpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    const Tensor& grad_output,
+    const Tensor& indices,
+    int64_t num_weights,
+    int64_t padding_idx,
+    bool scale_grad_by_freq) {
+  std::optional<Device> common_device = std::nullopt;
+  c10::impl::check_and_update_common_device(
+      common_device,
+      grad_output,
+      "xpu::embedding_dense_backward",
```
Comment (fengyuan14) [inline src/aten/Embedding.cpp:18]:
```suggestion
      "embedding_dense_backward_xpu",
```

File: src/aten/Embedding.cpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      "grad_output");
+  c10::impl::check_and_update_common_device(
+      common_device, indices, "xpu::embedding_dense_backward", "indices");
```
Comment (fengyuan14) [inline src/aten/Embedding.cpp:21]:
```suggestion
      common_device, indices, "embedding_dense_backward_xpu", "indices");
```

File: src/aten/Embedding.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  return native::xpu::embedding_dense_backward(
```
Comment (fengyuan14) [inline src/aten/Embedding.cpp:22]:
```suggestion
  return native::xpu::embedding_dense_backward_kernel(
```

File: src/aten/sycl/Embedding.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+namespace xpu {
+
+template <typename index_t>
+struct EmbeddingDenseBackwardEqFunctor {
+  auto operator()(index_t a, index_t b) const {
+    return a == b;
+  }
+};
+
+Tensor embedding_dense_backward(
```
Comment (fengyuan14) [inline src/aten/sycl/Embedding.cpp:18]:
```suggestion
Tensor embedding_dense_backward_kernel(
```

File: src/aten/sycl/Embedding.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include <ATen/native/TensorIterator.h>
+
+namespace at {
+namespace native {
+namespace xpu {
+
+Tensor embedding_dense_backward(
```
Comment (fengyuan14) [inline src/aten/sycl/Embedding.h:9]:
```suggestion
Tensor embedding_dense_backward_kernel(
```

File: src/aten/Embedding.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    const Tensor& grad_output,
+    const Tensor& indices,
+    int64_t num_weights,
+    int64_t padding_idx,
+    bool scale_grad_by_freq) {
+  std::optional<Device> common_device = std::nullopt;
+  c10::impl::check_and_update_common_device(
+      common_device,
+      grad_output,
+      "embedding_dense_backward_xpu",
```
Comment (fengyuan14) [inline src/aten/Embedding.cpp:19]:
```suggestion
      "xpu::embedding_dense_backward",
```

File: src/aten/Embedding.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+      "grad_output");
+  c10::impl::check_and_update_common_device(
+      common_device, indices, "embedding_dense_backward_xpu", "indices");
```
Comment (fengyuan14) [inline src/aten/Embedding.cpp:22]:
```suggestion
      common_device, indices, "xpu::embedding_dense_backward", "indices");
```

File: src/aten/XPUFallback.template [hunk 8]
```diff
@@ -223,7 +223,8 @@ TORCH_LIBRARY_IMPL(aten, XPU, m) {
     "_embedding_bag",
     "_embedding_bag_dense_backward",
     "_embedding_bag_per_sample_weights_backward",
-    "embedding_dense_backward",
+    "_embedding_bag_forward_only",
```
Comment (fengyuan14) [inline src/aten/XPUFallback.template:226]: Why we add a new fallback?

Comment (xytintel) [reply src/aten/XPUFallback.template:226]: Align with RegisterCUDA

**Comment by xytintel** [comment]
test/xpu/nn/test_embedding_xpu.py::TestEmbeddingNNDeviceTypeXPU::test_embedding_dense_grad_xpu PASSED

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**